Add section about clearing the sessions store during OMERO.web upgrade (#2312)
openmicroscopy/ome-documentation: omero/sysadmins/omeroweb-upgrade.rst

You should remove all previously installed plugins and install the latest
versions using pip.

.. _clearing_session_store:

Clear the sessions store (optional)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

During an upgrade, it might be necessary to clear the sessions store
before restarting OMERO.web. This is the case when either OMERO.web or
Django introduced backwards incompatible changes in the way sessions are
stored or if the session serialization format has been modified via
:property:`omero.web.session_serializer`.

.. warning::

   Clearing the sessions store will terminate all OMERO.web sessions and log
   out all users.

The process for clearing the store depends on the storage backend configured
via :property:`omero.web.session_engine`:

- if :property:`omero.web.session_engine` is either unset or set to
  `omeroweb.filesessionstore` or `django.contrib.sessions.backends.file`,
  OMERO.web will use file-based sessions. The session files are stored under a

>   temporary folder determined by the the output of `tempfile.gettempdir()`,

Reviewer: ```suggestion
  temporary folder determined by the output of `tempfile.gettempdir()`,
```

  usually `/tmp` and prefixed either by the value of

>   :property:`omero.web.session_cookie_name` if the property is set of by

Reviewer 2: of -> or

>   :property:`omero.web.session_cookie_name` if the property is set of by

Author: ```suggestion
  :property:`omero.web.session_cookie_name` if the property is set or by
```

>   :property:`omero.web.session_cookie_name` if the property if set of by

Reviewer: ```suggestion
  :property:`omero.web.session_cookie_name` if the property is set of by
```

>   :property:`omero.web.session_cookie_name` if the property if set of by

Reviewer 2: typos: ...if the property *is* set *or* by... - Also duplicate `with` in previous line

  `sessionid` by default.

>   For a default configuration, the the following command will

Reviewer 2: duplicate `the`

  typically delete all file sessions::

      $ rm /tmp/sessionid*

- if :property:`omero.web.session_engine` is set to
  `django.contrib.sessions.backends.cache`, OMERO.web uses cached sessions
  with a cache backend configured via :property:`omero.web.caches`. For a

>   Redis cache backend, sessions are stored using keys prefixed with

Reviewer: Add link to Redis?

Author: Happy to do add a link. While writing this PR, I realised we had no reference section about the support of Redis as a backend (it's mostly mentioned as part of the installation workflow). Are you thinking of linking to upstream Redis? or django-redis?

Related this discussion, this PR makes the authoritative assumption that files and Redis cache are the primary supported backends for storing sessions. Looking into the code, there are other backends that may be configured including database-backed sessions or Memcached cache https://docs.djangoproject.com/en/3.2/topics/http/sessions/. A potential follow-up of this PR might be to expand on the list of session engines and describe their level of support.

versions using pip.

.. _clearing_session_store:

Clear the sessions store (optional)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

During an upgrade, it might be necessary to clear the sessions store
before restarting OMERO.web. This is the case when either OMERO.web or
Django introduced backwards incompatible changes in the way sessions are
stored or if the session serialization format has been modified via
:property:`omero.web.session_serializer`.

.. warning::

   Clearing the sessions store will terminate all OMERO.web sessions and log
   out all users.

The process for clearing the store depends on the storage backend configured
via :property:`omero.web.session_engine`:

- if :property:`omero.web.session_engine` is either unset or set to
  `omeroweb.filesessionstore` or `django.contrib.sessions.backends.file`,
  OMERO.web will use file-based sessions. The session files are stored under a
  temporary folder determined by the output of `tempfile.gettempdir()`,
  usually `/tmp` and prefixed either by the value of
  :property:`omero.web.session_cookie_name` if the property is set or by
  `sessionid` if the property is unset.

>   For a default configuration, the following command will

Reviewer: maybe indicate earlier that by default it is unset
something like ``` :property:`omero.web.session_engine` (default: unset)```

Author: This will be true also if `omero.web.session_engine` is explicitly set to a file backend.

By default configuration, I was rather thinking to the expectation that `SESSION_COOKIE_NAME` and possibly other Django envvars would not be overriden. Maybe I'll invert the two sentence here i.e. first give the command that should work in 95% of the cases then mention the variables that could modify the location of the temporary sessions

Reviewer: The way it is written at the moment made one think that the default is unset.
Swapping the order should avoid the confusion

You should remove all previously installed plugins and install the latest
versions using pip.

.. _clearing_session_store:

Clear the sessions store (optional)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

During an upgrade, it might be necessary to clear the sessions store
before restarting OMERO.web. This is the case when either OMERO.web or
Django introduced backwards incompatible changes in the way sessions are
stored or if the session serialization format has been modified via
:property:`omero.web.session_serializer`.

.. warning::

   Clearing the sessions store will terminate all OMERO.web sessions and log
   out all users.

The process for clearing the session store depends on the storage backend:

> - OMERO.web session data is stored on the filesystem if the

Reviewer: data is plural. We have a mix of "data is" and "data are" in our doc.
We will need to fix that

Author: Oh dear, I didn't mean to enter such an unchartered territory. To my defense, the Django documentation is extensively use of `session data` with the singular form of the verb. 😄 
I think I'll just use `sessions` instead of `session data` everywhere.

  :property:`omero.web.session_engine` property is set to either
  `omeroweb.filesessionstore` or  `django.contrib.sessions.backends.file`
  or if :property:`omero.web.session_engine` is unset.

  Session data is stored under a temporary folder determined by the output of
  `tempfile.gettempdir()`, usually `/tmp`. By default, each session is
  stored as a separate file prefixed with `sessionid` so the following command
  will delete all stored sessions::

    $ rm /tmp/sessionid*

  If :property:`omero.web.session_cookie_name` is set, its value will be
  used as the prefix of the file sessions and the command above needs
  to be modified accordingly.

- OMERO.web session data is stored using the `Redis <https://redis.io/>`_ store
  if the :property:`omero.web.session_engine` property is set to
  `django.contrib.sessions.backends.cache` and Redis is configured via the
  :property:`omero.web.caches` property by setting the `BACKEND` to

>   'django.core.cache.backends.redis.RedisCache' and `LOCATION` to

Reviewer: ```suggestion
  `django.core.cache.backends.redis.RedisCache` and `LOCATION` to
```

  used as the prefix of the file sessions and the command above needs
  to be modified accordingly.

- OMERO.web sessions are stored using the `Redis <https://redis.io/>`_ store
  if the :property:`omero.web.session_engine` property is set to
  `django.contrib.sessions.backends.cache` and Redis is configured via the
  :property:`omero.web.caches` property by setting the `BACKEND` to
  `django.core.cache.backends.redis.RedisCache` and `LOCATION` to
  the URL of the Redis instance.

>   Session data is stored as key/value pairs in the Redis store with the name of

Reviewer: ``Sessions are stored as key/value`` should also be adjusted to ``Sessions are..`` in that case.
Same for ``Session data is stored under a temporary folder```
